File: fsharp/raindrops/Raindrops.fs
module Raindrops

    let private dictionary:(int * string) array = [|(3, "Pling"); (5, "Plang"); (7, "Plong")|]

    let private appender (number:int) (buffer:string) (factor:int, word:string) =
        match number % factor with
        | 0 -> buffer + word
        | _ -> buffer

    let private sentenceCreator (number:int):string =
        dictionary
        |> Array.fold (appender number) ""

    let convert (number:int):string =
        let sentence:string = sentenceCreator number
        match sentence with
        | "" ->
            number
            |> string
        | _ -> sentence

File: solutions/scala/raindrops/1/src/main/scala/Raindrops.scala
object Raindrops
{

  private def dictionary:Array[(Int, String)] = Array((3, "Pling"), (5, "Plang"), (7, "Plong"))

  private def appender(number:Int, buffer:String, factor:Int, word:String) =
    number % factor match
    case 0 => buffer + word
    case _ => buffer

  private def sentenceCreator (number:Int):String = dictionary.foldLeft[String]("")((buffer:String, entry:(Int, String)) => appender(number, buffer, entry._1, entry._2))

  def convert(n: Int): String =
  {
    val sentence:String = sentenceCreator(n)
    sentence match
    case "" => n.toString()
    case _ => sentence
  }
}

